This porject we will work with a dataset of used cars from ebay Kleinanzeigen, a classfields section of the German eBay website. The aim of this project is to clean the data and analyze the included used car listing.

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv('autos.csv',encoding='Latin-1')
autos[:10]

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [3]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We can find the dataset contains 20 columns, most of which are strings. some columns have null values, but none have more than ~20% null values and the column names use camelcase instead of Python's preferred snakecase. Hence we need to convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [5]:
maps = {'yearOfRegistration':'registration_year',
       'monthOfRegistration':'registration_month',
       'notRepairedDamage':'unrepaired_damage',
       'dateCreated':'ad_created',
       'dateCrawled':'date_crawled',
       'offerType':'offer_type',
       'vehicleType':'vehicle_type',
       'powerPS':'power_p_s',
       'fuelType':'fuel_type',
       'nrOfPictures':'nr_of_pictures',
       'postalCode':'postal_code',
       'lastSeen':'last_seen'}
autos = autos.rename(columns=maps)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 16:37:21  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From above, we can find that the price and odometer columns are numeric values stored as text. Hence we need to convert them to int datatype and rename the column to odometer_km

In [7]:
autos['price']=autos['price'].str.replace('$','').str.replace(',','').astype(int)


In [8]:
autos['odometer']=autos['odometer'].str.replace('km','').str.replace(',','').astype(int)

In [9]:
autos[:10]

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   
5  privat    Angebot   7900     test          bus               2006   
6  privat    Angebot    300     test    limousine               1995   
7  privat    Angebot   1990  control    limousine               1998   
8  privat    Angebot    250     test          NaN               2000   
9  privat    Angebot    590  control          bus               1997   

     gearbox  power_p_s    model  odometer  registration_month fuel_type  \
0    manuell        158   andere    150000                   3       lpg   
1  automatik        286      7er    150000                   6    benzin   
2    manuell        102     golf     70000                   7    benzin   
3  automatik         71   fortwo     70000                   6    benzin   
4    manuell          0    focus    150000                   7    benzin   
5  automatik        150  voyager    150000                   4    diesel   
6    manuell         90     golf    150000                   8    benzin   
7    manuell         90     golf    150000                  12    diesel   
8    manuell          0    arosa    150000                  10       NaN   
9    manuell         90   megane    150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   
5    chrysler               NaN  2016-03-21 00:00:00               0   
6  volkswagen               NaN  2016-03-20 00:00:00               0   
7  volkswagen              nein  2016-03-16 00:00:00               0   
8        seat              nein  2016-03-22 00:00:00               0   
9     renault              nein  2016-03-16 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50  
5        22962  2016-04-06 09:45:21  
6        31535  2016-03-23 02:48:59  
7        53474  2016-04-07 03:17:32  
8         7426  2016-03-26 18:18:10  
9        15749  2016-04-06 10:46:35

In [10]:
autos = autos.rename(columns={'odometer':'odometer_km'})

In [11]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [12]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [13]:
autos['price'].unique().shape

(2357,)

In [14]:
autos['odometer_km'].unique().shape

(13,)

In [15]:
autos['price'].value_counts().head()

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64

In [16]:
autos['odometer_km'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [17]:
autos['date_crawled'].str[:10].value_counts(normalize=True,
                                           dropna=False).sort_index()

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [18]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

From above we found that the registration_year column contains some odd values. The minimum 1000 is before cars were invented. The max is 9999, many years into the future.

In [19]:
brands = autos['brand'].unique()

In [20]:
brand={}
for i in brands:
    isthebrand = autos['brand'] == i
    mean_price = autos[isthebrand]['price'].mean()
    brand[i] = mean_price
brand

{'alfa_romeo': 3943.562310030395,
 'audi': 8965.560354891431,
 'bmw': 8252.918953766808,
 'chevrolet': 6432.929328621908,
 'chrysler': 3286.0552486187844,
 'citroen': 42657.46362339515,
 'dacia': 5897.736434108527,
 'daewoo': 1038.3544303797469,
 'daihatsu': 1552.09375,
 'fiat': 12134.20642201835,
 'ford': 7105.662546708824,
 'honda': 3889.8596491228072,
 'hyundai': 5316.754098360656,
 'jaguar': 11076.506493506493,
 'jeep': 11363.209090909091,
 'kia': 5707.3258426966295,
 'lada': 2476.9032258064517,
 'lancia': 3070.3508771929824,
 'land_rover': 18934.272727272728,
 'mazda': 3962.542932628798,
 'mercedes_benz': 29511.955428812842,
 'mini': 10392.393867924528,
 'mitsubishi': 3314.0492610837437,
 'nissan': 4588.879310344828,
 'opel': 5106.092657022524,
 'peugeot': 3010.8688186813188,
 'porsche': 44537.97959183674,
 'renault': 2351.301996672213,
 'rover': 1494.5217391304348,
 'saab': 3143.7,
 'seat': 4219.431455897981,
 'skoda': 6305.044529262086,
 'smart': 3482.971469329529,
 'sonstige_au

In [21]:
mileage={}
for i in brands:
    isthebrand = autos['brand'] == i
    mean_price = autos[isthebrand]['odometer_km'].mean()
    mileage[i] = mean_price
mileage

{'alfa_romeo': 131109.4224924012,
 'audi': 129643.9411627364,
 'bmw': 132521.64302818198,
 'chevrolet': 99522.96819787986,
 'chrysler': 133149.17127071825,
 'citroen': 119764.62196861627,
 'dacia': 84728.68217054264,
 'daewoo': 121708.86075949368,
 'daihatsu': 114843.75,
 'fiat': 117037.4617737003,
 'ford': 124131.93446392642,
 'honda': 123709.27318295739,
 'hyundai': 106782.7868852459,
 'jaguar': 121298.7012987013,
 'jeep': 126409.09090909091,
 'kia': 112640.44943820225,
 'lada': 86774.19354838709,
 'lancia': 123157.8947368421,
 'land_rover': 118333.33333333333,
 'mazda': 125132.10039630119,
 'mercedes_benz': 130886.14279678918,
 'mini': 89375.0,
 'mitsubishi': 126293.10344827586,
 'nissan': 118978.7798408488,
 'opel': 129298.66324848929,
 'peugeot': 127352.33516483517,
 'porsche': 97363.94557823129,
 'renault': 128223.79367720465,
 'rover': 136449.27536231885,
 'saab': 143750.0,
 'seat': 122061.63655685441,
 'skoda': 110947.83715012722,
 'smart': 100756.06276747503,
 'sonstige_autos'

In [22]:
brand_series = pd.Series(brand)
mileage_series = pd.Series(mileage)
price_df = pd.DataFrame(brand_series,columns=['mean_price'])
mileage_df = pd.DataFrame(mileage_series,columns=['mean_mileage'])

In [23]:
price_df

mean_price
alfa_romeo       3943.562310
audi             8965.560355
bmw              8252.918954
chevrolet        6432.929329
chrysler         3286.055249
citroen         42657.463623
dacia            5897.736434
daewoo           1038.354430
daihatsu         1552.093750
fiat            12134.206422
ford             7105.662547
honda            3889.859649
hyundai          5316.754098
jaguar          11076.506494
jeep            11363.209091
kia              5707.325843
lada             2476.903226
lancia           3070.350877
land_rover      18934.272727
mazda            3962.542933
mercedes_benz   29511.955429
mini            10392.393868
mitsubishi       3314.049261
nissan           4588.879310
opel             5106.092657
peugeot          3010.868819
porsche         44537.979592
renault          2351.301997
rover            1494.521739
saab             3143.700000
seat             4219.431456
skoda            6305.044529
smart            3482.971469
sonstige_autos  38300.840659
subaru           3769.981651
suzuki           4038.952218
toyota           5097.941653
trabant          1543.525641
volkswagen       6384.167400
volvo           31689.908096